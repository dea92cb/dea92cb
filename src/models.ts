// Extracted from https://github.com/gitfrosh/lotr-api/tree/release/backend/models

export type Book = {
  _id: string;
  name: string;
}

export type BookChapter = {
  chapterName: string;
}

export type Chapter = {
  _id: string;
  chapterName: string;
  book: string;
  bookName: string;
  ChapterData: string;
}

export type Character = {
  _id: string;
  name: string;
  wikiUrl: string;
  race: string;
  birth?: string | number | null;
  gender?: string;
  death?: string | number;
  hair?: string;
  height?: string;
  realm?: string;
  spouse?: string;
}

export type Movie = {
  _id: string;
  name: string;
  runtimeInMinutes: number;
  budgetInMillions: number;
  boxOfficeRevenueInMillions: number;
  academyAwardNominations: number;
  academyAwardWins: number;
  rottenTomatoesScore: number;
}

export type Quote = {
  _id: string;
  dialog: string;
  movie: string;
  character: string;
}