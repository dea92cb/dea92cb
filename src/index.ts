import { URLSearchParams } from 'url';
import fetch from 'cross-fetch';
import { Book, Chapter, Character, Movie, Quote } from './models';

export interface PaginationOptions {
  pagination?: {
    limit?: number;
    page?: number;
    offset?: number;
  };
}

export interface SortOptions<T> {
  sort?: {
    direction: 'asc' | 'desc';
    key: Extract<keyof T, string>;
  };
}

export type ListQueryOptions<T> = PaginationOptions & SortOptions<T> & FilterOptions

export interface FilterOptions {
  additionalOptions?: string;
}

export class LotrClient {
  public book = {
    list: (options?: ListQueryOptions<Book>) => this.get<Book, Book[]>('/book', options),
    get: (id: string) => this.get<Book, Book>(`/book/${id}`),
    listChapters: (id: string, options: ListQueryOptions<Chapter>) => this.get<Chapter, Chapter[]>(`/book/${id}/chapter`, options),
  };
  public movie = {
    list: (options?: ListQueryOptions<Movie>) => this.get<Movie, Movie[]>('/movie', options),
    get: (id: string) => this.get<Movie, Movie>(`/movie/${id}`),
    listQuotes: (id: string, options: ListQueryOptions<Quote>) => this.get<Quote, Quote[]>(`/movie/${id}/quotes`, options),
  };
  public character = {
    list: (options?: ListQueryOptions<Character>) => this.get<Character, Character[]>('/character', options),
    get: (id: string) => this.get<Character, Character>(`/character/${id}`),
    listQuotes: (id: string, options: ListQueryOptions<Quote>) => this.get<Quote, Quote[]>(`/character/${id}/quote`, options),
  };
  public quote = {
    list: (options?: ListQueryOptions<Quote>) => this.get<Quote, Quote[]>('/quote', options),
    get: (id: string) => this.get<Quote, Quote>(`/quote/${id}`),
  };
  public chapter = {
    list: (options?: ListQueryOptions<Chapter>) => this.get<Chapter, Chapter[]>('/chapter', options),
    get: (id: string) => this.get<Chapter, Chapter>(`/chapter/${id}`),
  };
  constructor(private options: {apiKey: string}) {
  }
  private constructQuery<T>(options: ListQueryOptions<T>): string {
    const searchParams = new URLSearchParams();

    if (options.pagination) {
      Object.entries(options.pagination).forEach(([option, value]) => searchParams.set(option, String(value)));
    }

    if (options.sort) {
      searchParams.set('sort', `${options.sort.key}:${options.sort.direction}`);
    }

    let qs = searchParams.toString();

    if (options.additionalOptions) {
      qs += '&' + options.additionalOptions;
    }

    return qs;
  }

  private async get<T, R>(path: string, options: ListQueryOptions<T> = {}): Promise<R> {
    const headers = {
      Accept: 'application/json',
      Authorization: `Bearer ${this.options.apiKey}`,
    };
    const query = this.constructQuery(options);
    const rawData = await fetch(`https://the-one-api.dev/v2/${path}${query}`, {
      headers: headers,
    });
    return rawData.json();
  }
}