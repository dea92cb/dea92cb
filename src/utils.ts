export async function retryLoop<T>(
  pF: () => Promise<T>,
  number: number,
  failMessage: string,
): Promise<T> {
  let e: any;
  for (let i = 0; i < number; i += 1) {
    try {
      const result = await pF();
      if (result) {
        return result;
      }
    } catch (err) {
      e = err;
      await new Promise(resolve => setTimeout(resolve, 50 * (2 ** i) + 50 * (2**i) * Math.random()));
    }
  }
  if (e) {
    throw new Error(`${failMessage}: ${e.toString()}`);
  }
  throw new Error(failMessage);
}