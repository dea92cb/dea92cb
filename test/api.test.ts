import { LotrClient } from '../src';

test('look up book list', async () => {
  const client = new LotrClient({ apiKey: '' });
  expect(await client.book.list()).toMatchSnapshot();
});

test('look up book by id', async () => {
  const client = new LotrClient({ apiKey: '' });
  expect(await client.book.get('5cf5805fb53e011a64671582')).toMatchSnapshot();
});