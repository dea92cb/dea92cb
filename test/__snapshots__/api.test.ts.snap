// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`look up book by id 1`] = `
Object {
  "docs": Array [
    Object {
      "_id": "5cf5805fb53e011a64671582",
      "name": "The Fellowship Of The Ring",
    },
  ],
  "limit": 1000,
  "offset": 0,
  "page": 1,
  "pages": 1,
  "total": 1,
}
`;

exports[`unauthenticated URL works 1`] = `
Object {
  "docs": Array [
    Object {
      "_id": "5cf5805fb53e011a64671582",
      "name": "The Fellowship Of The Ring",
    },
    Object {
      "_id": "5cf58077b53e011a64671583",
      "name": "The Two Towers",
    },
    Object {
      "_id": "5cf58080b53e011a64671584",
      "name": "The Return Of The King",
    },
  ],
  "limit": 1000,
  "offset": 0,
  "page": 1,
  "pages": 1,
  "total": 3,
}
`;
